### PR TITLE
Experimental map search

### PR DESCRIPTION
**Pull request by jakozaur**
Experimental feature that lets search `Map(String, String)`.

It generates SQL where like:
```
(arrayExists((x) -> x ILIKE '%key%', mapKeys("attributes")) OR arrayExists((x) -> x ILIKE '%key%', mapValues("attributes")))
```

Limitations:
1. Only full-text is supported.
2. Only `Map(String, String)`.
3. No highlighter.
Screenshot 2024-08-01 at 19 20 13

### CODE REVIEW

**Comment by nablaone** [comment]
Field/column with type `Map(String, String)` is now multifield. It supports suffixes:
- '.key'  which is an array of keys
- `.value' which is an array of values
- `.text` both of them 

Screenshot 2024-08-22 at 16 56 10

**Comment by jakozaur** [comment]
@nablaone how about plural `keys`, `values`?

**Comment by nablaone** [comment]
Clickhouse has a special syntax for  maps:


https://clickhouse.com/docs/en/sql-reference/data-types/map#reading-subcolumns-of-map

:)

File: quesma/model/expr_string_renderer.go [hunk 1]
```diff
@@ -21,8 +21,11 @@ func AsString(expr Expr) string {
 }
 
 func (v *renderer) VisitColumnRef(e ColumnRef) interface{} {
+	// TODO this should be done as the last step in the pipeline, not here
 	name := strings.TrimSuffix(e.ColumnName, ".keyword")
-	name = strings.TrimSuffix(name, "::keyword") // Not sure if this is the best place to do this
+	name = strings.TrimSuffix(name, "::keyword")
+	name = strings.TrimSuffix(name, ".key")
+	name = strings.TrimSuffix(name, ".value")
```
Comment (jakozaur) [inline quesma/model/expr_string_renderer.go:28]: Maybe constants?

File: quesma/quesma/functionality/field_capabilities/field_caps.go [hunk 2]
```diff
@@ -69,6 +69,10 @@ func handleFieldCapsIndex(cfg config.QuesmaConfiguration, schemaRegistry schema.
 					addFieldCapabilityFromSchemaRegistry(fields, fmt.Sprintf("%s.keyword", fieldName.AsString()), schema.TypeKeyword, resolvedIndex)
 				case "keyword":
 					addFieldCapabilityFromSchemaRegistry(fields, fmt.Sprintf("%s.text", fieldName.AsString()), schema.TypeText, resolvedIndex)
+				case "map":
+					addFieldCapabilityFromSchemaRegistry(fields, fmt.Sprintf("%s.text", fieldName.AsString()), schema.TypeText, resolvedIndex)
+					addFieldCapabilityFromSchemaRegistry(fields, fmt.Sprintf("%s.key", fieldName.AsString()), schema.TypeText, resolvedIndex)
+					addFieldCapabilityFromSchemaRegistry(fields, fmt.Sprintf("%s.value", fieldName.AsString()), schema.TypeText, resolvedIndex)
```
Comment (jakozaur) [inline quesma/quesma/functionality/field_capabilities/field_caps.go:75]: I would like `.keys` and `.values` suffix more.

File: quesma/quesma/schema_map_transformation.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+				return existsInKey
+
+			case (op == "ILIKE" || op == "LIKE") && scope == scopeValues:
+
+				variableName := "x"
+				lambda := model.NewLambdaExpr([]string{variableName}, model.NewInfixExpr(model.NewLiteral(variableName), op, e.Right.Accept(b).(model.Expr)))
+				existsInValue := model.NewFunction("arrayExists", lambda, model.NewFunction("mapValues", left))
+
+				return existsInValue
```
Comment (jakozaur) [inline quesma/quesma/schema_map_transformation.go:115]: Feels really redundant, it can be abstract to something common.

**Review by jakozaur — COMMENTED** [review]
Looks good, but would rethink keys and values.

File: quesma/model/expr_string_renderer.go (hunk 1 above)
Comment (nablaone) [reply quesma/model/expr_string_renderer.go]: Added.

**Review by jakozaur — COMMENTED** [review]
Good for me, but can't approve as I started it.

**Review by nablaone — APPROVED** [review]

**Pull request merged**